### PR TITLE
RH6: Solving the Build Error by adding kfree_rcu for RHEL <6.4

### PR DESCRIPTION
**Pull request by mboitei**

### CODE REVIEW

File: hv-rhel6.x/hv/vmbus_drv.c [hunk 1]
```diff
@@ -87,6 +87,50 @@ struct hv_device_info {
...
 
+#if defined(RHEL_RELEASE_VERSION) && (RHEL_RELEASE_CODE < RHEL_RELEASE_VERSION(6,4))
+
+#define __is_kfree_rcu_offset(offset) ((offset) < 4096)
+
+/*
+ * Helper macro for kfree_rcu() to prevent argument-expansion eyestrain.
+ */
+#define __kfree_rcu(head, offset) \
+	do { \
```
Comment (santoshx) [inline hv-rhel6.x/hv/vmbus_drv.c]: I suppose symbol kfree_rcu doesn't exist in old kernel hence you adding it.
Did you try to find how similar operation done without this symbol?

Comment (mboitei) [reply hv-rhel6.x/hv/vmbus_drv.c]: I have modified git commit message. kfree_rcu was introduced by the commit c6100fd. 
The kfree_rcu is a Kernel symbol for newer Kernels and does not present in older kernel, so added the kfree_rcu symbol in LIS to handle older Kernels build issues. 
I did look into for a similar operation in older kernel as that could have needed to rewrite the vmbus_chan_release().

**Comment by mboitei** [comment]
We can change only in LIS, not is Kernel code. That's why LIS code modified for kfree_rcu() to support RHEL Kernel < 6.4.

**Comment by santoshx** [comment]
>           We can change only in LIS, not is Kernel code. That's why LIS code modified for kfree_rcu() to support RHEL Kernel < 6.4.

I'm aware that we can change only in LIS code and not in kernel. 
Does this mean you are sure the part in kernel is not needed?

**Comment by mboitei** [comment]
> es this mean you are sure the part in kernel is not needed



> > ```
> >       We can change only in LIS, not is Kernel code. That's why LIS code modified for kfree_rcu() to support RHEL Kernel < 6.4.
> > ```
> 
> I'm aware that we can change only in LIS code and not in kernel.
> Does this mean you are sure the part in kernel is not needed?

The fix only considers the kfree-rcu, so other functions were not taken.

**Comment by mboitei** [comment]
@stevecui Can you please review the code with respect to the original patch https://lwn.net/Articles/433493/ ?

**Comment by santoshx** [comment]
> > es this mean you are sure the part in kernel is not needed
> 
> > > ```
> > >       We can change only in LIS, not is Kernel code. That's why LIS code modified for kfree_rcu() to support RHEL Kernel < 6.4.
> > > ```
> > 
> > 
> > I'm aware that we can change only in LIS code and not in kernel.
> > Does this mean you are sure the part in kernel is not needed?
> 
> The fix only considers the kfree-rcu, so other functions were not taken.

Have you analyzed and found out that changes made to other functions are not needed for kfree_rcu to work correctly?

**Comment by mboitei** [comment]
> > > es this mean you are sure the part in kernel is not needed
> > 
> > 
> > > > ```
> > > >       We can change only in LIS, not is Kernel code. That's why LIS code modified for kfree_rcu() to support RHEL Kernel < 6.4.
> > > > ```
> > > 
> > > 
> > > I'm aware that we can change only in LIS code and not in kernel.
> > > Does this mean you are sure the part in kernel is not needed?
> > 
> > 
> > The fix only considers the kfree-rcu, so other functions were not taken.
> 
> Have you analyzed and found out that changes made to other functions are not needed for kfree_rcu to work correctly?

I have analyzed that I do not need the other functions.

**Comment by santoshx** [comment]
> > > > es this mean you are sure the part in kernel is not needed
> > > 
> > > 
> > > > > ```
> > > > >       We can change only in LIS, not is Kernel code. That's why LIS code modified for kfree_rcu() to support RHEL Kernel < 6.4.
> > > > > ```
> > > > 
> > > > 
> > > > I'm aware that we can change only in LIS code and not in kernel.
> > > > Does this mean you are sure the part in kernel is not needed?
> > > 
> > > 
> > > The fix only considers the kfree-rcu, so other functions were not taken.
> > 
> > 
> > Have you analyzed and found out that changes made to other functions are not needed for kfree_rcu to work correctly?
> 
> I have analyzed that I do not need the other functions.

Do you mean functionality of kfree_rcu is not dependent on changes made to those kernel functions when it was introduced?

**Comment by mboitei** [comment]
> > > > > es this mean you are sure the part in kernel is not needed
> > > > 
> > > > 
> > > > > > ```
> > > > > >       We can change only in LIS, not is Kernel code. That's why LIS code modified for kfree_rcu() to support RHEL Kernel < 6.4.
> > > > > > ```
> > > > > 
> > > > > 
> > > > > I'm aware that we can change only in LIS code and not in kernel.
> > > > > Does this mean you are sure the part in kernel is not needed?
> > > > 
> > > > 
> > > > The fix only considers the kfree-rcu, so other functions were not taken.
> > > 
> > > 
> > > Have you analyzed and found out that changes made to other functions are not needed for kfree_rcu to work correctly?
> > 
> > 
> > I have analyzed that I do not need the other functions.
> 
> Do you mean functionality of kfree_rcu is not dependent on changes made to those kernel functions when it was introduced?

In this case I am using the kfree_rcu to freeing an object and accordingly modified the code. The  functions/symbols  related to kfree_rcu added for <RHEL 6.4.  I am not using the __rcu_reclaim, so the function is not added for < 6.4.

**Comment by santoshx** [comment]
> > > > > > es this mean you are sure the part in kernel is not needed
> > > > > 
> > > > > 
> > > > > > > ```
> > > > > > >       We can change only in LIS, not is Kernel code. That's why LIS code modified for kfree_rcu() to support RHEL Kernel < 6.4.
> > > > > > > ```
> > > > > > 
> > > > > > 
> > > > > > I'm aware that we can change only in LIS code and not in kernel.
> > > > > > Does this mean you are sure the part in kernel is not needed?
> > > > > 
> > > > > 
> > > > > The fix only considers the kfree-rcu, so other functions were not taken.
> > > > 
> > > > 
> > > > Have you analyzed and found out that changes made to other functions are not needed for kfree_rcu to work correctly?
> > > 
> > > 
> > > I have analyzed that I do not need the other functions.
> > 
> > 
> > Do you mean functionality of kfree_rcu is not dependent on changes made to those kernel functions when it was introduced?
> 
> In this case I am using the kfree_rcu to freeing an object and accordingly modified the code. The functions/symbols related to kfree_rcu added for <RHEL 6.4. I am not using the __rcu_reclaim, so the function is not added for < 6.4.

It appears to me, reclaim is needed after kfree_rcu. Please check.

**Comment by mboitei** [comment]
Code modified and rpm build log copied.
RedHat RHEL 6.9 latest] Copied : .\report\report_PROVISION.xml --> .\report\report_PROVISION-33118-junit.xml
[RedHat RHEL 6.9 latest] Analysing results..
[RedHat RHEL 6.9 latest] PASS  : 1
[RedHat RHEL 6.9 latest] FAIL  : 0
[RedHat RHEL 6.9 latest] ABORT : 0
[RedHat RHEL 6.9 latest] Copying all files to original working directory.
[RedHat RHEL 6.9 latest] Cleaning C:\Windows\TEMP\az\33118
[RedHat RHEL 6.9 latest] Setting workspace to original location: C:\Jenkins-Slaves\azure-slave-2-executor-5\workspace\pipeline-lis-rpm-build
[RedHat RHEL 6.9 latest] Exiting with code : 0

**Comment by mboitei** [comment]
All the comments are addressed, so merging the code.

**Pull request merged**